### ISSUE TITLE
Pruned definition of class template function causes re-declaration error

### ISSUE THREAD

**Issue by psalz** [illuhad/hipSYCL#16]
In this example

```c++
class foo {
    template<typename T>
    void bar(T value);
};

template <typename T>
void baz(T) {}

template <typename T>
void foo::bar(T value) {
    baz(value);
}

int main() {
    return 0;
}
```

`nvcc` will throw a re-declaration error of `foo::bar`, as only the declaration is left after pruning. I think the correct approach in this case would be to comment out the entire definition outside the class.

**Comment by illuhad** [comment]
We could indeed comment out the entire definition, or we replace the definition with a trivial one i.e. `{}`, potentially with dummy return value. I'll think about what the best solution will be, in any case I think this will be easy to fix :)

**Comment by psalz** [comment]
I think you're right, and I think replacing pruned definitions with `{}` might be a good approach in general (not just in this case). I suppose the return value doesn't matter, as compilers probably won't do any static analysis on uninstantiated templates.

**Comment by illuhad** [comment]
It would also simplify the code slightly since for constructors, we are already inserting `{}` to avoid having an initializer list in a constructor that has been turned into a declaration (which causes nvcc to crash with some internal assertion failure).
The reason why I went the route of commenting out definitions was that clang's replace function would always crash - but in hindsight it could be that this was just because of the weird begin/end source locations... I'll try that again later :)

**Comment by illuhad** [comment]
Here are my current findings:
* Even when inserting `{}`, we still need to comment out the existing code block instead of just overwriting it, since removing several code ranges in the same rewrite buffer apparently can really screw up the rewrite buffer, leading to a garbled result. Perhaps it works better if we fill the replacement buffer with enough spaces so that it is at least of equal size as the original code block?
* gcc actually warns about missing return statements in uninstantiated templates, so a dummy return value is in fact necessary
* Dummy return values require several special cases, for example in pruned lambda functions clang defines the return value of type `auto`, so we cannot just do something like `return *reinterpret_cast<return_type*>(new int);` which I was working with. Perhaps it would be possible to just not prune lambda functions since it's sufficient to prune the enclosing function, but there are other problems, for example it seems clang typedefs `bool` internally as `_Bool` (causing `hipsycl_transform_source` to insert `_Bool` return values) which nvcc doesn't know

Perhaps it is simpler to just comment out the entire definition as before, with special treatment for class template functions.

**Comment by psalz** [comment]
I see, in that case it might be easier to comment out the entire definition, yes.

I'd like to help figure this out but I'm absolutely swamped right now (I actually have a bunch of things regarding hipSYCL that I'd like to discuss in my backlog -- hopefully I can get around filing some issues in the next couple of days).

**Comment by illuhad** [comment]
No problem, I also did not have so much time this week :) If you have time at some point, please try PR #17 with your code. This PR should fix your input case and still allow for correct compilation of SYCL parallel STL code.

**Comment by illuhad** [comment]
(all known occurences of this issue have been reported to be fixed by PR #28, reopen if this occurs again)
